### ISSUE TITLE
Add v1beta1 AlertmanagerConfig CRD

### ISSUE THREAD

**Issue by simonpasquier** [prometheus-operator/prometheus-operator#4677]
The AlertmanagerConfig CRD has been introduced as v1alpha1 in v0.43.0, one and half years ago. We've collected great feedback so far and the CRD now supports all Alertmanager configuration options. We'd like more users to benefit from it and offering a v1beta1 version of the CRD would increase adoption (beta version means that the project won't remove the API in the future and provides increased support).

I want to use this issue as a way to share and track the work towards this goal.

## Overview

Though alpha API versions may be removed in any release without prior deprecation notice, we want the migration from v1alpha1 to v1beta1 to be easy and seamless. The project will implement conversion webhooks to convert v1alpha1 objects to v1beta1 (and vice-versa). Initially the CRD will continue to use v1alpha1 as the stored version (e.g. the representation persisted in etcd) but to encourage users to switch to the new API version, it will switch in a couple of releases to v1beta1.

# Changes

## API changes

* Remove the `regex` field from the [Matcher](https://prometheus-operator.dev/docs/operator/api/#matcher) type.
  * When converting from v1alpha1 to v1beta1
    * "empty" matchType + false regex turns into `matchType: "="`
    * "empty" matchType + true regex turns into matchType: "=~"
  * When converting from v1beta1 to v1alpha1
    * Nothing changes
* Anticipate support for active time intervals (see #4656)
  * Rename `spec.muteTimeIntervals` to `spec.timeIntervals`
    * When converting from v1alpha1 to v1beta1
      * Copy `spec.muteTimeIntervals` to `spec.timeIntervals`
    * When converting from v1beta1 to v1alpha1
      * Copy `spec.timeIntervals` to `spec.muteTimeIntervals`
  * Rename the MuteTimeInterval type to TimeInterval
  * Rename the TimeInterval type to TimePeriod

## Improvements

* Push field validations to the open API specification instead of when reconciling the resources.

## Supported Alertmanager versions

We continue to support all Alertmanager versions without limitation. 

## Deprecation plan

* [v0.57.0] -> [v0.58.0] (~3 months): Alertmanager CRD v1beta1 is opt-in (e.g. not included in `bundle.yaml`).
* [v0.59.0] -> [v0.62.0] (~6 months): Alertmanager CRD v1beta1 is deployed by default, v1alpha1 marked as deprecated but still defined as the stored version and used by the operator.
* [v0.63.0]: Alertmanager v1alpha1 is removed, the operator switches using v1beta1.

## Documentation

* Detailed documentation about how users can migrate their objects to the new version.
* Clear documentation of the deprecation period and future removal.

## kube-prometheus

* Automate the deployment of the conversion webhook.

## Questions

* ~~Should we take opportunity to move away from `coreos.monitoring.com` and use `prometheus-operator.dev`?~~
 ** This would add unnecessary complexity for the users.
* Should we offer tools to convert YAML manifests from `v1alpha1` to `v1beta1` (like `cmctl convert` from [cert-manager](https://cert-manager.io/docs/usage/cmctl/#convert))?
 ** We will determine later if there's a need for it but given the deprecation plan, it's not a priority for now.

## Links

[Kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)
[CRD versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/)

**Comment by simonpasquier** [comment]
@prometheus-operator/prometheus-operator-reviewers please have a look at this :)

**Comment by simonpasquier** [comment]
> Should we take opportunity to move away from coreos.monitoring.com and use prometheus-operator.dev?

I would be in favor of this but I'm not 100% sure that it wouldn't complicate the conversion (need to be checked).

> Should we offer tools to convert YAML manifests from v1alpha1 to v1beta1 (like cmctl convert from [cert-manager](https://cert-manager.io/docs/usage/cmctl/#convert))?

Given that the API changes should be rather limited, I don't think it is worth the effort (or at least it doesn't have to exist before releasing v1beta1).

**Comment by slashpai** [comment]
> Should we take opportunity to move away from coreos.monitoring.com and use prometheus-operator.dev?

Other resources in `v1` are `coreos.monitoring.com` so if we change this only for alertmanager would that look odd? :)

**Comment by PhilipGough** [comment]
> if we change this only for alertmanager would that look odd?

I tend to agree that it would. Although the original group is legacy at this point, it is still well known and I am not sure what we really gain by moving away from it?

**Comment by simonpasquier** [comment]
@slashpai this is a good point and having to deal with a new API version + a new group at the same time would probably complicate things too much (both for us and and for users). Lets keep `monitoring.coreos.com` for now!

**Comment by simonpasquier** [comment]
> > if we change this only for alertmanager would that look odd?
> 
> I tend to agree that it would. Although the original group is legacy at this point, it is still well known and I am not sure what we really gain by moving away from it?

The main reason would be consistency and more clarity. But yes, the (user benefit / potential trouble) ratio might not be very good :)

**Comment by slashpai** [comment]
This looks good to be closed as https://github.com/prometheus-operator/prometheus-operator/pull/4709 merged

**Comment by jkroepke** [comment]
Hi @simonpasquier 

We have a request to integrate v1beta1 into kube-prometheus-stack.

https://github.com/prometheus-community/helm-charts/pull/3938

Since the schuedle of the v1beta seems not up2date: 

Is from Prometheus Operator project point of view the v1beta1 API safe to use? Or still in preview? I'm asking here because the v1beta1 is only part of the full CRDs.

Is the converational webhook manadory/prerequirement? The PR author says, that it works without.
